crates/consensus/src/transaction/eip4844/builder.rs

```diff
@@ -0,0 +1,408 @@
+    }
+}
+
+/// A strategy for coding and decoding data into sidecars. Coder instances are
+/// responsible for encoding and decoding data into and from the sidecar. They
+/// are called by the [`SidecarBuilder`] during the [`ingest`],
+/// [`take`], and (if `c_kzg` feature enabled) `build` methods.
+///
+/// This trait allows different downstream users to use different bit-packing
+/// strategies. For example, a simple coder might only use the last 31 bytes of
+/// each blob, while a more complex coder might use a more sophisticated
+/// strategy to pack data into the low 6 bits of the top byte.
+///
+/// [`ingest`]: SidecarBuilder::ingest
+/// [`take`]: SidecarBuilder::take
+pub trait SidecarCoder {
+    /// Calculate the number of field elements required to store the given
+    /// data.
+    fn required_fe(&self, data: &[u8]) -> usize;
+
+    /// Code a slice of data into the builder.
+    fn code(&mut self, builder: &mut PartialSidecar, data: &[u8]);
+
+    /// Finish the sidecar, and commit to the data. This method should empty
+    /// any buffer or scratch space in the coder, and is called by
+    /// [`SidecarBuilder`]'s `take` and `build` methods.
+    fn finish(self, builder: &mut PartialSidecar);
+
+    /// Decode all slices of data from the blobs.
+    fn decode_all(&mut self, blobs: &[Blob]) -> Option<Vec<Vec<u8>>>;
+}
+
+/// Simple coder that only uses the last 31 bytes of each blob. This is the
+/// default coder for the [`SidecarBuilder`].
+///
+/// # Note
+///
+/// Because this coder sacrifices around 3% of total sidecar space, we do not
+/// recommend its use in production. It is provided for convenience and
+/// non-prod environments.
+///
+/// # Behavior
+///
+/// This coder encodes data as follows:
+/// - The first byte of every 32-byte word is empty.
+/// - Data is pre-pended with a 64-bit big-endian length prefix, which is right padded with zeros to
+///   form a complete word.
+/// - The rest of the data is packed into the remaining 31 bytes of each word.
+/// - If the data is not a multiple of 31 bytes, the last word is right-padded with zeros.
+///
+/// This means that the following regions cannot be used to store data, and are
+/// considered "wasted":
+///
+/// - The first byte of every 32-byte word.
+/// - The right padding on the header word containing the data length.
+/// - Any right padding on the last word for each piece of data.
+#[derive(Debug, Copy, Clone, Default)]
+pub struct SimpleCoder;
+
+impl SimpleCoder {
+    /// Decode an some bytes from an iterator of valid FEs.
+    ///
+    /// Returns `Ok(Some(data))` if there is some data.
+    /// Returns `Ok(None)` if there is no data (length prefix is 0).
+    /// Returns `Err(())` if there is an error.
+    fn decode_one<'a>(mut fes: impl Iterator<Item = WholeFe<'a>>) -> Result<Option<Vec<u8>>, ()> {
+        let first = fes.next().ok_or(())?;
+        let mut num_bytes = u64::from_be_bytes(first.as_ref()[1..9].try_into().unwrap()) as usize;
```

Comment:
would be nice to have a method on `WholeFe` that does the array ref conversion, documented with assumptions about how `WholeFe` is constructed